internal/catalog/internal/mappers/nodemapper/node_mapper.go

```diff
@@ -0,0 +1,100 @@
+package nodemapper
+
+import (
+	"context"
+	"sync"
+
+	"github.com/hashicorp/consul/internal/catalog/internal/types"
+	"github.com/hashicorp/consul/internal/controller"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"google.golang.org/protobuf/proto"
+)
+
+type NodeMapper struct {
+	lock             sync.Mutex
+	nodesToWorkloads map[string][]controller.Request
+}
+
+func New() *NodeMapper {
+	return &NodeMapper{
+		nodesToWorkloads: make(map[string][]controller.Request),
+	}
+}
+
+func (m *NodeMapper) NodeIDFromWorkload(workload *pbresource.Resource, workloadData *pbcatalog.Workload) *pbresource.ID {
+	return &pbresource.ID{
+		Type:    types.NodeV1Alpha1Type,
+		Tenancy: workload.Id.Tenancy,
+		Name:    workloadData.NodeName,
+	}
+}
+
+func (m *NodeMapper) MapNodeToWorkloads(_ context.Context, _ controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return m.nodesToWorkloads[res.Id.Name], nil
+}
+
+func (m *NodeMapper) TrackWorkload(workloadID *pbresource.ID, nodeID *pbresource.ID) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	reqs, ok := m.nodesToWorkloads[nodeID.Name]
+	if ok {
+		for _, req := range reqs {
+			// if the workload already is mapped to the node
+			if proto.Equal(req.ID, workloadID) {
+				return
+			}
+		}
+	}
+
+	// Check if this workload is being tracked for another node and
+	// remove the link. This would only occur if the workloads
+	// associated node name is changed.
+	m.removeWorkloadTrackingLocked(workloadID)
+
+	// Now set up the latest tracking
+	m.nodesToWorkloads[nodeID.Name] = append(reqs, controller.Request{ID: workloadID})
+}
+
+func (m *NodeMapper) RemoveWorkloadTracking(workloadID *pbresource.ID) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.removeWorkloadTrackingLocked(workloadID)
+}
+
+func (m *NodeMapper) removeWorkloadTrackingLocked(workloadID *pbresource.ID) {
+	// TODO make this not perform in O(<num global workloads>) time
```

Comment:
🤔 It looks like in some cases (when a workload is added, deleted or when the node name has changed), this would still need to go through all workloads in the worst case in the reconcile (if we consider the number of nodes negligible here). If so, then how are we improving it from the "naive" implementation where we list all workloads with a certain node name in the dependency mapper every time a node changes? 

There's not a good way to compare the number of events in the two cases. In the naive approach, this would be any change to a node (including its health), which arguably doesn't happen that often. For workloads, if it's a dynamic environment like k8s, workloads could come and go frequently maybe even more frequently than node health would change. Maybe you have a better insight into how to think about this.

You're mentioning in the TODO that we should improve this to not go through all workloads. How do you think we can do that?

Reply:
I think what you pointed out was that this implementation was actually worse than the naive implementation. The naive one would have done a list when mapping node events to workloads to construct the list of workload events. This was going to iterate over all node associated workloads every time a workload was reconciled (which is going to be far more often then node events firing.

So I just fixed it. Keeping a map of currently tracked workloads means that checking if a workload is already checked and determining if the node association has changed can be done in constant time (its a map index + string equality).